File: android/app/src/main/java/app/threadmind/domain/ActionCard.kt
package app.threadmind.domain

enum class ActionType { CREATE_MEETING, CREATE_CONTACT, UPDATE_CONTACT }
enum class ActionStatus { DRAFT, BLOCKED, READY, CONFIRMED, EXECUTING, SUCCEEDED, FAILED, CANCELLED }

data class EvidenceRef(
    val sourceId: String,
    val messageId: String?,
    val excerpt: String,
    val confidence: Double,
)

data class ActionCard(
    val id: String,
    val submissionId: String,
    val type: ActionType,
    val version: Int,
    val fields: Map<String, String>,
    val evidence: List<EvidenceRef>,
    val fieldConfidence: Map<String, Double>,
    val validationIssues: List<String>,
    val targetAccountId: String?,
    val status: ActionStatus,
    val blockers: List<String>,
    val confirmedSnapshot: ConfirmedActionSnapshot? = null,
)

data class ConfirmedActionSnapshot(
    val actionCardId: String,
    val type: ActionType,
    val version: Int,
    val fields: Map<String, String>,
    val evidence: List<EvidenceRef>,
    val targetAccountId: String,
    val idempotencyKey: String,
)

object ActionCardPolicy {
    private val required = mapOf(
        ActionType.CREATE_MEETING to setOf("title", "startsAt", "endsAt", "timezone", "targetCalendarId"),
        ActionType.CREATE_CONTACT to setOf("displayName", "contactMethod", "targetContactAccountId"),
        ActionType.UPDATE_CONTACT to setOf("targetContactId", "changes"),
    )

    fun evaluate(card: ActionCard): ActionCard {
        val blockers = card.validationIssues.map { "validation:$it" }.toMutableList()
        blockers += required.getValue(card.type)
            .filter { card.fields[it].isNullOrBlank() }
            .map { "missing:$it" }
        if (card.evidence.isEmpty()) blockers += "missing:evidence"
        if (card.targetAccountId.isNullOrBlank()) blockers += "missing:targetAccountId"
        return card.copy(status = if (blockers.isEmpty()) ActionStatus.READY else ActionStatus.BLOCKED, blockers = blockers)
    }

    fun confirm(card: ActionCard): ActionCard {
        val ready = evaluate(card)
        require(ready.status == ActionStatus.READY) { "Only a ready card can be confirmed" }
        val target = requireNotNull(ready.targetAccountId)
        val snapshot = ConfirmedActionSnapshot(
            actionCardId = ready.id,
            type = ready.type,
            version = ready.version,
            fields = ready.fields.toMap(),
            evidence = ready.evidence.toList(),
            targetAccountId = target,
            idempotencyKey = "${ready.id}:${ready.version}:$target",
        )
        return ready.copy(status = ActionStatus.CONFIRMED, confirmedSnapshot = snapshot)
    }

    fun edit(
        card: ActionCard,
        fields: Map<String, String>,
        resolvedValidationIssues: Set<String> = emptySet(),
    ): ActionCard {
        require(resolvedValidationIssues.all(card.validationIssues::contains)) { "Only current validation issues can be resolved" }
        return evaluate(card.copy(
            version = card.version + 1,
            fields = fields.toMap(),
            fieldConfidence = fields.mapValues { (field, value) ->
                if (card.fields[field] == value) card.fieldConfidence[field] ?: 1.0 else 1.0
            },
            validationIssues = card.validationIssues.filterNot(resolvedValidationIssues::contains),
            status = ActionStatus.DRAFT,
            confirmedSnapshot = null,
        ))
    }
}

File: server/src/domain/action-card.ts
import { createHash, randomUUID } from "node:crypto";
import { isDeepStrictEqual } from "node:util";
import { invariant } from "./errors.ts";
import type { ActionCard, ActionReceipt, ConfirmedActionSnapshot, EvidenceRef } from "./model.ts";

const requiredFields = {
  create_meeting: ["title", "startsAt", "endsAt", "timezone", "targetCalendarId"],
  create_contact: ["displayName", "contactMethod", "targetContactAccountId"],
  update_contact: ["targetContactId", "changes"],
} as const;

export function evaluateCard(card: ActionCard): ActionCard {
  const missing = requiredFields[card.type].filter((field) => {
    const value = card.fields[field];
    return value === undefined || value === null || value === "";
  });
  const blockers = [
    ...card.validationIssues.map((issue) => `validation:${issue}`),
    ...missing.map((field) => `missing:${field}`),
  ];
  if (card.evidence.length === 0) blockers.push("missing:evidence");
  if (!card.targetAccountId) blockers.push("missing:targetAccountId");
  return { ...card, blockers, status: blockers.length === 0 ? "ready" : "blocked" };
}

export function editCard(
  card: ActionCard,
  fields: Record<string, unknown>,
  evidence: EvidenceRef[] = card.evidence,
  resolvedValidationIssues: string[] = [],
  targetAccountId: string | undefined = card.targetAccountId,
): ActionCard {
  invariant(!["executing", "succeeded", "cancelled"].includes(card.status), "card_not_editable", "Card can no longer be edited");
  const unknownIssue = resolvedValidationIssues.find((issue) => !card.validationIssues.includes(issue));
  invariant(!unknownIssue, "unknown_validation_issue", "Only current validation issues can be resolved");
  const { confirmedSnapshot: _snapshot, confirmedAt: _confirmedAt, ...editable } = card;
  const fieldConfidence = Object.fromEntries(Object.entries(fields).map(([field, value]) => [
    field,
    Object.hasOwn(card.fields, field) && isDeepStrictEqual(card.fields[field], value)
      ? card.fieldConfidence[field] ?? 1
      : 1,
  ]));
  return evaluateCard({
    ...editable,
    version: card.version + 1,
    fields: structuredClone(fields),
    evidence: structuredClone(evidence),
    fieldConfidence,
    validationIssues: card.validationIssues.filter((issue) => !resolvedValidationIssues.includes(issue)),
    ...(targetAccountId ? { targetAccountId } : {}),
    status: "draft",
  });
}

export function confirmCard(card: ActionCard, now = new Date()): ActionCard {
  if (card.status === "confirmed" && card.confirmedSnapshot?.version === card.version) return structuredClone(card);
  const evaluated = evaluateCard(card);
  invariant(evaluated.status === "ready", "card_not_ready", `Card is blocked: ${evaluated.blockers.join(", ")}`);
  const digest = createHash("sha256")
    .update(`${card.accountId}:${card.id}:${card.version}:${card.targetAccountId}`)
    .digest("hex");
  const snapshot: ConfirmedActionSnapshot = Object.freeze({
    actionCardId: card.id,
    accountId: card.accountId,
    type: card.type,
    version: card.version,
    fields: Object.freeze(structuredClone(card.fields)),
    targetAccountId: card.targetAccountId!,
    evidence: Object.freeze(structuredClone(card.evidence)),
    idempotencyKey: digest,
  });
  return { ...evaluated, status: "confirmed", confirmedSnapshot: snapshot, confirmedAt: now.toISOString() };
}

export function recordExecution(
  card: ActionCard,
  result: { status: "succeeded"; targetRecordId: string } | { status: "failed" | "cancelled"; errorCode?: string; errorMessage?: string },
  previousAttempts: ActionReceipt[],
  now = new Date(),
  receiptId: string = randomUUID(),
): { card: ActionCard; receipt: ActionReceipt } {
  invariant(card.status === "confirmed" || card.status === "failed", "card_not_confirmed", "Only a confirmed snapshot can execute");
  invariant(card.confirmedSnapshot, "snapshot_missing", "Confirmed snapshot is required");
  const startedAt = now.toISOString();
  const provider = card.type === "create_meeting" ? "android_calendar" : "android_contacts";
  const receipt: ActionReceipt = {
    id: receiptId,
    accountId: card.accountId,
    actionCardId: card.id,
    confirmedVersion: card.confirmedSnapshot.version,
    attempt: previousAttempts.length + 1,
    status: result.status,
    provider,
    ...(result.status === "succeeded" ? { targetRecordId: result.targetRecordId } : {}),
    ...(result.status !== "succeeded" && result.errorCode ? { errorCode: result.errorCode } : {}),
    ...(result.status !== "succeeded" && result.errorMessage ? { errorMessage: result.errorMessage } : {}),
    startedAt,
    completedAt: now.toISOString(),
  };
  invariant(receipt.status !== "succeeded" || receipt.targetRecordId, "target_missing", "Successful execution needs a target record id");
  return { card: { ...card, status: result.status }, receipt };
}
